File: holoeasy-example-packetevents/src/main/java/org/holoeasy/plugin/ExamplePlugin.java
package org.holoeasy.plugin;


import com.github.retrooper.packetevents.PacketEvents;
import io.github.retrooper.packetevents.factory.spigot.SpigotPacketEventsBuilder;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.plugin.java.JavaPlugin;
import org.holoeasy.HoloEasy;
import org.holoeasy.builder.TextLineModifiers;
import org.holoeasy.hologram.Hologram;

import org.holoeasy.line.Line;
import org.holoeasy.packet.PacketImpl;
import org.holoeasy.pool.IHologramPool;
import org.holoeasy.reactive.MutableState;
import org.jetbrains.annotations.NotNull;


public class ExamplePlugin extends JavaPlugin {

    private HoloEasy holoEasy;

    @Override
    public void onLoad() {
        PacketEvents.setAPI(SpigotPacketEventsBuilder.build(this));
        //On Bukkit, calling this here is essential, hence the name "load"
        PacketEvents.getAPI().load();
    }

    @Override
    public void onDisable() {
        //Terminate the instance (clean up process)
        PacketEvents.getAPI().terminate();

        // Since 4.4.0: destroy pools
        holoEasy.destroyPools();
    }

    @Override
    public void onEnable() {
        //Initialize!
        PacketEvents.getAPI().init();

        // ** Bind the library
        holoEasy = new HoloEasy(this, PacketImpl.PacketEvents);

        // ** Create a MyHolo Pool, why not?
        IHologramPool<MyHolo> myPool = holoEasy.startInteractivePool(60);

        getCommand("hologram").setExecutor((sender, cmd, s, args) -> {

            Player player = ((Player) sender);
            Location location = player.getLocation();


            // ** Add holo to myPool
            MyHolo hologram = new MyHolo(holoEasy, location);
            hologram.show(myPool);

            return true;
        });


        // ** Why not update all holograms 'status' item after 30 seconds?
        Bukkit.getScheduler().runTaskLaterAsynchronously(this, () -> {


            for (MyHolo hologram : myPool.getHolograms()) {

                // ** Updates the line
                hologram.status.update(new ItemStack(Material.GREEN_DYE));
            }

        }, 20L * 30);
    }



    public static class MyHolo extends Hologram {

        private final MutableState<Integer> clickCount = mutableStateOf(0); // can be any type

        public Line<String> counter = textLine(ChatColor.translateAlternateColorCodes('&', "&7Clicked {} times"), new TextLineModifiers()
                .args(clickCount)
                .clickable(player -> clickCount.update(it -> it + 1)));
        public Line<ItemStack> status = itemLine(new ItemStack(Material.RED_DYE));

        public MyHolo(@NotNull HoloEasy lib, @NotNull Location location) {
            super(lib, location);
        }

    }


}

File: holoeasy-example-protocollib/src/main/kotlin/ExamplePlugin.kt
package org.holoeasy.plugin

import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.plugin.Plugin
import org.bukkit.plugin.java.JavaPlugin
import org.holoeasy.HoloEasy
import org.holoeasy.builder.TextLineModifiers
import org.holoeasy.hologram.Hologram
import org.holoeasy.packet.PacketImpl

class ExamplePlugin : JavaPlugin() {

    override fun onEnable() {

        // ** Bind the library
        HoloEasy.bind(this, PacketImpl.ProtocolLib)

        getCommand("hologram")?.setExecutor { sender, _, _, _ ->

            val location = (sender as Player).location

            // ** Create a basic hologram and add on the Standard Pool
            val hologram = HelloWorldHologram(location)
            hologram.show()


            //  ** Serialize
            val serialized: Map<String, Any> = hologram.serialize()

            // ** Now you can remove it from the Standard Pool
            hologram.hide()

            // ** Deserialize the previous hologram
            val deserialized : HelloWorldHologram = Hologram.deserialize(serialized)

            // ** Show it
            deserialized.show()

            true
        }
    }


    class MyHolo(plugin: Plugin, location: Location) : Hologram(location) {

        private val clickCount = mutableStateOf(0) // can be any type

        val counter = textLine("Clicked {} times", TextLineModifiers()
            .args(clickCount)
            .clickable { _ -> clickCount.update { it + 1 } })

        var status= blockLine(ItemStack(Material.RED_DYE))

    }


}
